File: old/2018/cpp/01.cpp
// Advent of Code 2018, day 1
// (c) aichingert

#include <iostream>
#include <vector>
#include <numeric>
#include <functional>
#include <set>
#include <fstream>
#include <string>

int part1(const std::vector<int>& a) {
  return std::accumulate(a.begin(), a.end(), 0);
}

int part2(const std::vector<int>& a) {
  int cur = 0;
  int i = 0;
  std::set<int> b {};

  while (b.insert(cur).second) {
    cur += a[i];
    i = (i + 1) % a.size();
  }

  return cur;
}

int main(void) {
  std::ifstream ifs("../input/01");
  std::vector<int> a {};
  std::string line {};

  while (std::getline(ifs, line)) {
    int idk = std::stoi(line);
    a.push_back(idk);
  }

  std::cout << "Part 1: "<< part1(a) << "\n";
  std::cout << "Part 2: "<< part2(a) << "\n";
}

File: advent_rainbow/2015/haskell/src/01.hs
-- Advent of Code 2015, day 1
-- (c) aichingert

module Main where

import Data.Char

part1 :: [Char] -> Int -> Int
part1 [] loc = loc
part1 (x:xs) loc | x == '(' = part1 xs (loc + 1)
                 | x == ')' = part1 xs (loc - 1)

part2 :: [Char] -> Int -> Int -> Int
part2 [] loc cur | loc < 0 = cur 
                 | otherwise = -1
part2 (x:xs) loc cur | loc < 0 = cur
                     | x == '(' = part2 xs (loc + 1) (cur + 1)
                     | x == ')' = part2 xs (loc - 1) (cur + 1)

main :: IO ()
main = do
  input <- readFile "input/01"
  putStrLn ("Part 1: " ++ (show (part1 input 0)))
  putStrLn ("Part 2: " ++ (show (part2 input 0 0)))
